### PR TITLE
feat: Hookdeck webhook signature verification

### PR DESCRIPTION
**Pull request by leggetter**
Adds the ability to verify the Hookdeck webhook signature.

```ts
const headers = request.headers;
const rawBody = await request.text();
const result = await verifyWebhookSignature({
    headers,
    rawBody,
    signingSecret: process.env.HOOKDECK_SIGNING_SECRET,
})

if(!result.isValidSignature) {
    // Reject the webhook payload
}
else {
    // Proceed with the webhook payload
}
```

Also check that Hookdeck verified the initial request from the Source.

```ts
const headers = request.headers;
const rawBody = await request.text();
const result = await verifyWebhookSignature({
    headers,
    rawBody,
    signingSecret: process.env.HOOKDECK_SIGNING_SECRET,
    config: {
        checkSourceVerification: true,
    }
})

if(!result.isValidSignature) {
    // Reject the webhook payload
}
else {
    // Proceed with the webhook payload
}
```

### CODE REVIEW

File: src/webhooks/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ *     }
+ *     else {
+ *         // Proceed with the webhook payload
+ *     }
+ */
+export const verifyWebhookSignature = async ({
+    signature,
+    rawBody,
+    signingSecret,
+    secondarySignture,
```
Comment (alexbouchardd) [inline src/webhooks/helpers.ts]: Typo

File: src/webhooks/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}: HookdeckVerifyArguments): Promise<HookdeckVerificationResult> => {
+    const signatureCheck = await createHS256Signature({ signingSecret, data: rawBody });
+
+    return {
+        isValidSignature: signatureCheck === signature || signatureCheck === secondarySignture,
```
Comment (alexbouchardd) [inline src/webhooks/helpers.ts]: Need to decide how to handle `x-hookdeck-verified`, a flag to the method maybe?

File: src/webhooks/helpers.ts (hunk 1 above)
Comment (leggetter) [reply src/webhooks/helpers.ts]: 👍 fixed in 5d33f9592a72e759528a716f941705be040e1b50

File: src/webhooks/helpers.ts (hunk 2 above)
Comment (leggetter) [reply src/webhooks/helpers.ts]: What do you think of the DX of making the `verifyWebhookSignature` function require a key/value parameter of `headers`? That way the helper can extract the required headers and the calling of the function is more about telling it how to behave.

```ts
const result =
  await verifyWebhookSignature(
    {
      headers,
      rawBody,
      signingSignature,
      config: {
        checkSourceVerification: true, // true: check for x-hookdeck-verified and fails if it's false
      }
    }
  );
```

Comment (leggetter) [reply src/webhooks/helpers.ts]: Here's a full example with the two options:

```ts
import { NextResponse } from "next/server";
import { verifyWebhookSignature } from "@hookdeck/sdk/webhooks/helpers";

export async function POST(request: Request) {

  // Option 1: explicitly pass in headers and other params
  const verificationResult = await verifyWebhookSignature({
    signature: request.headers.get("x-hookdeck-signature")!,
    secondarySignature: request.headers.get("x-hookdeck-signature-2"),
    sourceVerified: request.headers.get("x-hookdeck-verified"),
    rawBody: await request.text(),
    signingSecret: process.env.HOOKDECK_SIGNING_SECRET!,
  });

  // Option 2: developer needs to create header key/value pair
  // then pass in headers
  const headers: { [key: string]: string } = {};
  request.headers.forEach((value, key) => {
    headers[key] = value;
  });

  const verificationResult = await verifyWebhookSignature({
    headers,
    rawBody: await request.text(),
    signingSecret: process.env.HOOKDECK_SIGNING_SECRET!,
    config: { checkSourceVerification: true },
  });

  if (!verificationResult.isValidSignature) {
    return new NextResponse(
      JSON.stringify({ error: "Invalid Hookdeck signature" }),
      { status: 401 },
    );
  }
  const body = await request.json();
  console.log(body);
  return NextResponse.json({ received: true });
}
```

I think option 2 is a little more work to create the `headers` key-value pair. However, it removes the chances of people making errors when getting the header params and hides more of the implementation details.

Comment (alexbouchardd) [reply src/webhooks/helpers.ts]: Yeah I believe #2 is the most practical

> I think option 2 is a little more work to create the headers key-value pair. 

Not sure what you mean by that!

Comment (leggetter) [reply src/webhooks/helpers.ts]: > > I think option 2 is a little more work to create the headers key-value pair.
> 
> Not sure what you mean by that!

Getting the headers is a little more work for the developer using the SDK. Not more work to be done within the SDK.

File: src/webhooks/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
+// const { webcrypto, createHmac } = require("node:crypto");
+const { webcrypto } = require("node:crypto");
+const { subtle } = webcrypto;
+
+const ALGORITHM = { name: "HMAC", hash: "SHA-256" };
+
+const createHS256Signature = async ({ signingSecret, data }: { signingSecret: string; data: string }) => {
```
Comment (chadxz) [inline src/webhooks/helpers.ts]: What does “HS” stand for here?

File: src/webhooks/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
+// const { webcrypto, createHmac } = require("node:crypto");
```
Comment (chadxz) [inline src/webhooks/helpers.ts]: Remove commented code?

File: src/webhooks/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    return signature;
+};
+
+/**
+ * Error thrown when there are problems with the parameters used to verify webhooks.
+ */
+export class HookdeckWebhookVerificationParameterError extends Error {
+    constructor(message: string) {
+        super(message);
+        Object.setPrototypeOf(this, HookdeckWebhookVerificationParameterError.prototype);
```
Comment (chadxz) [inline src/webhooks/helpers.ts:23]: this.name = this.constructor.name is helpful too.

File: src/webhooks/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    headers,
+    rawBody,
+    signingSecret,
+    config = {},
+}: HookdeckWebhookVerificationArguments): Promise<HookdeckWebhookVerificationResult> => {
+    const signature = headers[HOOKDECK_HEADERS.SIGNATURE];
+    const secondarySignature = headers[HOOKDECK_HEADERS.SECONDARY_SIGNATURE];
+
+    if (!signature) {
+        throw new HookdeckWebhookVerificationParameterError(`The ${HOOKDECK_HEADERS.SIGNATURE} header is missing.`);
```
Comment (chadxz) [inline src/webhooks/helpers.ts]: Maybe quote the header name in the message like you do the last one

**Review by chadxz — APPROVED** [review]
LGTM. I wonder if the user may want this to never throw, though. In that case all the errors would be turned into like debug logs and instead return false for valid signature. Just a thought.

You may also want to check the shape of your error to make sure the name, stack trace etc look how you want them to look.

File: src/webhooks/helpers.ts (hunk 3 above)
Comment (alexbouchardd) [reply src/webhooks/helpers.ts]: Seems like a typo for `SHA`

**Pull request merged**